### ISSUE TITLE
AttributeRepository does not execute _afterSave() on Attribute resource model

### ISSUE THREAD

**Issue by lbajsarowicz** [magento/magento2#14181]
### Preconditions

1. Magento 2.2.3
2. PHP 7.1

### Steps to reproduce

1. Create Customer EAV attribute (with `system => 0` and `user_defined => 0` attributes) using `\Magento\Customer\Setup\CustomerSetup->addAttribute()`
2. Get Customer EAV attribute with `\Magento\Eav\Model\AttributeRepository->get()`
3. Set `$attribute->setData('used_in_forms', ['adminhtml_customer']);`
4. Execute `$repository->save()` using the `$attribute`.

### Expected result

1. As far as I see - we should have row for this attribute in `customer_form_attribute`
2. Xdebug should stop on breakpoint on `\Magento\Eav\Model\ResourceModel\Attribute::_afterSave` during install

### Actual result

1. There is no entry for new attribute in `customer_form_attribute`
2. Xdebug breaks on breakpoint at `\Magento\Eav\Model\ResourceModel\Entity\Attribute::_afterSave`

### Another approaches
* I don't use $attribute->save() for purpose. It's deprecated in `> 2.1`

**Comment by lbajsarowicz** [comment]
Discussed the issue with @Vinai and he recommended to inject with Dependency Injection the Attribute ResourceModel and then use `$resourceModel->save($attribute)`.

This works fine, but I'm not sure about this approach, because IMHO I should not use *Resource Model* to save `used_in_forms` especially when it's implemented in Magento as `_afterSave` for Attribute Resource Model.

**Comment by orlangur** [comment]
Hi @lbajsarowicz,

> 3. Set `$attribute->setData('used_in_forms', ['adminhtml_customer']);`

I don't see `setData` method in `\Magento\Eav\Api\Data\AttributeInterface`.

Please give more insight on expected result as what I see from implementation
```
public function save(\Magento\Eav\Api\Data\AttributeInterface $attribute)
    {
        try {
            $this->eavResource->save($attribute);
        } catch (\Exception $e) {
            throw new StateException(__('Cannot save attribute'));
        }
        return $attribute;
    }
```
it is quite universal while `\Magento\Eav\Model\ResourceModel\Entity\Attribute` is an abstract class used only in customer attribute.

**Comment by Vinai** [comment]
My understanding is that the interfaces provide no way to set the `used_in_forms` property without violating the interface contract (i.e. calling a method that is not declared on the interface).
When calling a method that is not part of the interface, it creates an implicit dependency on a concrete implementation.
Implicit dependencies cause issues such as static code analysis not picking up the real dependencies, alternative implementations of the interface but not the additional implicitly expected methods being incompatible, etc. In effect implicit dependencies negate all the benefits of dependency inversion (the D in SOLID).
Following this reasoning, implicit dependencies should be made explicit.
It follows further that instead of depending on the AttributeInterface and the repository, code setting the `used_in_forms` property should depend on the concrete attribute implementation and the resource model instead.

**Comment by lbajsarowicz** [comment]
@orlangur After getting into the issue bit deeper, I can agree that for Data Install script I can use specific Model (Customer Attribute) to set `used_in_forms` attribute and then use injected Resource Model to save this property for this specific attribute.

Probably I overacted to use Interface instead of Resource Model directly.

On the other hand - `getResourceModel()` and `$attribute->save()` is deprecated since `2.1` and lots of solutions across the Internet are misleading about this specific attribute (`used_in_forms`). That's why I wanted to find the "perfect" (looks like "too perfect") way.

To sum up - this issue can be closed as invalid.

**Comment by magento-engcom-team** [comment]
@lbajsarowicz Closing per your comment above.
Thanks!
